File: problems/Swift/563.二叉树的坡度.swift
/**
 * 563.二叉树的坡度 [Easy]
 * Time Complexity: O(n), Space Complexity: O(n)
 * https://leetcode-cn.com/problems/binary-tree-tilt/
 */

/**
 * Definition for a binary tree node.
 */
public class TreeNode {
  public var val: Int
  public var left: TreeNode?
  public var right: TreeNode?
  public init() { val = 0; left = nil; right = nil }
  public init(_ val: Int) { self.val = val; left = nil; right = nil }
  public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
    self.val = val
    self.left = left
    self.right = right
  }
}

class Solution {
  private var ans = 0

  func findTilt(_ root: TreeNode?) -> Int {
    dfs(root)
    return ans
  }

  private func dfs(_ root: TreeNode?) -> Int {
    guard let root = root else { return 0 }
    let left = dfs(root.left)
    let right = dfs(root.right)
    ans += abs(left - right)
    return left + right + root.val
  }
}

File: problems/rust-solutions/src/563.二叉树的坡度.rs
// 563.二叉树的坡度 [Easy]
// Time Complexity: O(n), Space Complexity: O(n)
// https://leetcode-cn.com/problems/binary-tree-tilt/
use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}

impl Solution {
  pub fn find_tilt(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    let ans = Solution::dfs(root);
    return ans.0;
  }

  fn dfs(root: Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
    if let Some(root) = root {
      let left = Solution::dfs(root.borrow().left.clone());
      let right = Solution::dfs(root.borrow().right.clone());
      return (
        left.0 + right.0 + (left.1 - right.1).abs(),
        left.1 + right.1 + root.borrow().val,
      );
    }

    return (0, 0);
  }
}
